### PR TITLE
feat: Added the parameter-free online learning algorithm FreeGrad

### PR DESCRIPTION
**Pull request by zmhammedi**
Added the parameter-free online learning algorithm FreeGrad from this paper: http://proceedings.mlr.press/v125/mhammedi20a/mhammedi20a.pdf

### CODE REVIEW

File: vowpalwabbit/freegrad.cc [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+{
+  auto FG = scoped_calloc_or_throw<freegrad>();
+  bool FreeGrad;
+  bool restart = false;
+  bool project = false;
+  bool adaptiveradius = true;
+  float radius; 
+
+  option_group_definition new_options("FreeGrad options");
+  new_options.add(make_option("freegrad", FreeGrad).keep().help("Diagonal FreeGrad Algorithm")).add(make_option("restart", restart).help("Use the FreeRange restarts"))
```
Comment (lalo) [inline vowpalwabbit/freegrad.cc:288]:
```suggestion
  new_options.add(make_option("freegrad", FreeGrad).necessary().keep().help("Diagonal FreeGrad Algorithm")).add(make_option("restart", restart).help("Use the FreeRange restarts"))
```

File: vowpalwabbit/freegrad.cc [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+      .add(make_option("project", project).help("Project the outputs to adapt to both the lipschitz and comparator norm")).add(make_option("radius", radius).help("Radius of the l2-ball for the projection. If not supplied, an adaptive radius will be used.")).add(make_option("fepsilon", FG->epsilon).default_value(1.f).help("Initial wealth"));
+
+  options.add_and_parse(new_options);
```
Comment (lalo) [inline vowpalwabbit/freegrad.cc:291]:
```suggestion
  options.add_parse_and_check_necessary(new_options);
```

File: vowpalwabbit/freegrad.cc [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+
+  if (!FreeGrad) { return nullptr; }
```
Comment (lalo) [inline vowpalwabbit/freegrad.cc:293]: if you commit the other two suggestions, you can remove this if

File: vowpalwabbit/freegrad.h [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#pragma once
```
Comment (olgavrou) [inline vowpalwabbit/freegrad.h:1]: file missing header?

File: vowpalwabbit/freegrad.cc [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+  auto learn_ptr = (fg_ptr->all->audit || fg_ptr->all->hash_inv) ? learn_freegrad<true> : learn_freegrad<false>;
+  auto* l = VW::LEARNER::make_base_learner(std::move(fg_ptr), learn_ptr, predict_ptr,
+      stack_builder.get_setupfn_name(VW::freegrad_setup), VW::prediction_type_t::scalar, VW::label_type_t::simple)
+                .set_learn_returns_prediction(true)
+                .set_params_per_weight(UINT64_ONE << stack_builder.get_all_pointer()->weights.stride_shift())
+                .set_save_load(save_load)
+                .set_end_pass(end_pass)
+                .build();
+
+  // (TODO) Check what the multipredict is about and set_and_pass
```
Comment (olgavrou) [inline vowpalwabbit/freegrad.cc:368]: can we remove this commented out code?

**Review by olgavrou — APPROVED** [review]

**Pull request merged**